File: tests/ut/py/test_worker/test_kernel_mode_entry.py
# Copyright (c) PyPTO Contributors.
# This program is free software, you can redistribute it and/or modify it under the terms and conditions of
# CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.
# -----------------------------------------------------------------------------------------------------------
# ruff: noqa: PLC0415
"""Hardware UT for the Python kernel-mode surface of ChipWorker.

The Python twin of tests/ut/cpp/hardware/test_kernel_mode_entry.cpp, and the
inverse of test_platform_comm.py's contract: there ChipWorker owns ACL bring-up
and stream lifetime internally, here the *caller* owns both and lends the stream
in. That inversion is what kernel mode is, so the test does its own device bind
and stream creation through ``_acl_bind_device`` / ``_acl_create_stream`` and
hands the resulting integer address to kernel_launch. In production that integer
comes from the framework instead — torch_npu.npu.current_stream().npu_stream.

TMR init and prepare succeed on private streams. Execution remains unsupported
until the binding/executor providers exist, so direct launch must reach the C ABI
and return INVALID_STATE. The caller stream must survive successful finalization.

Each case runs in a forked subprocess: kernel_init binds a runtime library into
the process and the ACL device bind is per-thread, so a fresh process per case
keeps one case's state out of the next.

The ``runtime`` marker is what makes conftest's resource phase dispatch these
rather than deselect them, so it is load-bearing rather than descriptive — the
runtime it names is the one whose binaries the cases load.
"""

from __future__ import annotations

import multiprocessing as mp
import tempfile
import traceback
from pathlib import Path

import pytest


def _finalize_worker(worker, result) -> None:
    if worker is not None:
        try:
            worker.finalize()
        except Exception as exc:  # noqa: BLE001
            result["ok"] = False
            result["finalize_error"] = str(exc)


def _run_case(case: str, device_id: int, platform: str, queue) -> None:
    """Subprocess body: stand up the caller's device + stream, then drive the
    kernel-mode surface and report what came back."""
    result: dict[str, object] = {"case": case, "stage": "start", "ok": False}
    stream = 0
    worker = None
    try:
        import _task_interface as native
        from simpler.task_interface import CallConfig, ChipCallable, ChipStorageTaskArgs, ChipWorker

        from simpler_setup.kernel_compiler import KernelCompiler
        from simpler_setup.runtime_builder import RuntimeBuilder

        bins = RuntimeBuilder(platform=platform).get_binaries("tensormap_and_ringbuffer", build=False)

        # The caller's device and stream. simpler must not create, reset or
        # destroy any of this.
        if case != "program_init_still_works":
            native._acl_bind_device(device_id)
            stream = native._acl_create_stream()
            result["stream_nonzero"] = bool(stream)
            result["stage"] = "borrowed"

        worker = ChipWorker()
        config = CallConfig()

        if case == "init_prepare_launch_close":
            worker.kernel_init(device_id, bins, config)
            assert worker._impl.initialized
            assert not worker.kernel_mode_supported
            with tempfile.TemporaryDirectory(prefix="kernel-entry-") as build_dir:
                binary = KernelCompiler(platform).compile_orchestration(
                    "tensormap_and_ringbuffer",
                    str(Path(__file__).resolve().parents[1] / "kernel_prepare_orchestration.cpp"),
                    build_dir=build_dir,
                )
            chip = ChipCallable.build(
                signature=[], func_name="kernel_prepare_orchestration", binary=binary, children=[]
            )
            callable_id = worker.kernel_prepare_callable(chip)
            assert callable_id == 0
            assert worker._callable_registry[callable_id] is chip
            with pytest.raises(RuntimeError, match="-1003") as excinfo:
                worker.kernel_launch(callable_id, ChipStorageTaskArgs(), stream)
            result["error"] = str(excinfo.value)
            worker.finalize()
            assert not worker._impl.initialized
            assert not worker._callable_registry
            native._acl_destroy_stream(stream)
            stream = 0
            result["stream_destroyed_by_caller"] = True
            result["ok"] = True

        elif case == "null_stream_rejected":
            with pytest.raises(ValueError, match="caller_stream") as excinfo:
                worker.kernel_launch(0, ChipStorageTaskArgs(), 0)
            result["error"] = str(excinfo.value)
            result["ok"] = "caller_stream" in str(excinfo.value)

        elif case == "generation_is_unique":
            first = native._ChipWorker.next_kernel_context_generation()
            second = native._ChipWorker.next_kernel_context_generation()
            result["first"], result["second"] = first, second
            result["ok"] = first != 0 and second > first

        elif case == "uninitialized_surface_refuses":
            errors = []
            for label, fn in (
                ("supported", lambda: worker.kernel_mode_supported),
                ("launch", lambda: worker.kernel_launch(0, ChipStorageTaskArgs(), stream)),
            ):
                try:
                    fn()
                except Exception as exc:  # noqa: BLE001
                    errors.append(label)
                    result[f"err_{label}"] = str(exc)
            result["refused"] = errors
            result["ok"] = errors == ["supported", "launch"]

        elif case == "program_init_still_works":
            # The program path must be unchanged by the kernel additions, and
            # the two identities must stay mutually exclusive on one worker.
            worker.init(device_id=device_id, bins=bins)
            result["stage"] = "program_init"
            result["initialized"] = bool(worker._impl.initialized)
            result["kernel_supported"] = bool(worker.kernel_mode_supported)
            try:
                worker.kernel_init(device_id, bins, config)
                result["second_init_refused"] = False
            except RuntimeError:
                result["second_init_refused"] = True
            worker.finalize()
            result["ok"] = result["initialized"] and not result["kernel_supported"] and result["second_init_refused"]

        else:
            raise AssertionError(f"unknown case {case}")

        result["stage"] = "done"
    except BaseException as exc:  # noqa: BLE001
        result["error"] = f"{type(exc).__name__}: {exc}"
        result["traceback"] = traceback.format_exc()
    finally:
        _finalize_worker(worker, result)
        if stream:
            try:
                import _task_interface as native

                native._acl_destroy_stream(stream)
                result["stream_destroyed_by_caller"] = True
            except Exception as exc:  # noqa: BLE001
                result["ok"] = False
                result["stream_destroy_error"] = str(exc)
        queue.put(result)


def _run_in_subprocess(case: str, device_id: int, platform: str) -> dict:
    ctx = mp.get_context("fork")
    queue = ctx.Queue()
    proc = ctx.Process(target=_run_case, args=(case, device_id, platform, queue))
    proc.start()
    proc.join(timeout=300)
    assert proc.exitcode is not None, f"case {case} did not exit within 300s"
    assert not queue.empty(), f"case {case} produced no result (exitcode={proc.exitcode})"
    return queue.get()


@pytest.mark.requires_hardware
@pytest.mark.platforms(["a2a3"])
@pytest.mark.device_count(1)
@pytest.mark.runtime("tensormap_and_ringbuffer")
@pytest.mark.parametrize(
    "case",
    [
        "init_prepare_launch_close",
        "null_stream_rejected",
        "generation_is_unique",
        "uninitialized_surface_refuses",
        "program_init_still_works",
    ],
)
def test_kernel_mode_surface_on_borrowed_stream(case, st_platform, st_device_ids):
    """Drive one kernel-mode case against a stream the test itself owns."""
    assert st_device_ids, "device_count(1) fixture must yield at least one id"
    result = _run_in_subprocess(case, int(st_device_ids[0]), st_platform)
    assert result["ok"], f"case {case} failed: {result}"


@pytest.mark.requires_hardware
@pytest.mark.platforms(["a2a3"])
@pytest.mark.device_count(1)
@pytest.mark.runtime("tensormap_and_ringbuffer")
def test_borrowed_stream_survives_kernel_finalize(st_platform, st_device_ids):
    """Successful finalization releases only the context-owned resources."""
    assert st_device_ids
    result = _run_in_subprocess("init_prepare_launch_close", int(st_device_ids[0]), st_platform)
    assert result.get("stream_nonzero"), f"test never obtained a stream: {result}"
    assert result["ok"], f"kernel lifecycle did not behave per contract: {result}"
    assert result.get("stream_destroyed_by_caller"), f"finalize invalidated the borrowed stream: {result}"

File: tests/ut/cpp/hardware/test_kernel_mode_entry.cpp
/*
 * Copyright (c) PyPTO Contributors.
 * This program is free software, you can redistribute it and/or modify it under the terms and conditions of
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 * -----------------------------------------------------------------------------------------------------------
 */

/*
 * Hardware UT for the kernel-mode entry layer on a2a3 onboard: the four C ABI
 * entries reached through a real libhost_runtime.so, on a device and stream the
 * test owns rather than simpler.
 *
 * The borrowing is the point. Every case here does its own aclInit +
 * aclrtSetDevice + aclrtCreateStream and hands that stream in, so what is under
 * test is the borrowed-device shape a kernel-mode caller presents. A test that
 * let simpler stand the device up would exercise the program-mode shape and
 * pass for the wrong reason.
 *
 * TMR initialization establishes its private streams without borrowing the
 * caller stream. Execution remains unsupported until the binding/executor
 * providers exist: supported is zero and launch returns INVALID_STATE.
 * Malformed arguments return INTERNAL before state admission. Finalization
 * releases the private context without invalidating the caller stream.
 *
 * Hardware classification: requires_hardware_a2a3 (ctest label) + CMake gate
 * SIMPLER_ENABLE_HARDWARE_TESTS. Device allocation is driven by CTest
 * RESOURCE_GROUPS + --resource-spec-file.
 *
 * PTO_HOST_RUNTIME_LIB_PATH and the three binary paths are baked in at
 * configure time by tests/ut/cpp/CMakeLists.txt.
 */

#include <dlfcn.h>

#include <cstdint>
#include <cstdlib>
#include <cstring>
#include <filesystem>
#include <fstream>
#include <iterator>
#include <string>
#include <vector>

#include <gtest/gtest.h>

#include "acl/acl.h"
#include "call_config.h"
#include "callable.h"
#include "callable_protocol.h"
// ChipWorker holds a unique_ptr<ChipRunLane>, whose deleter needs the complete
// type wherever a ChipWorker is destroyed.
#include "chip_run_lane.h"
#include "chip_worker.h"
#include "runtime_c_api.h"
#include "task_args.h"

namespace {

/// The four kernel-mode entries, resolved out of libhost_runtime.so the way
/// ChipWorker resolves them.
struct KernelModeApi {
    void *(*create_device_context)();
    void (*destroy_device_context)(void *);
    decltype(&simpler_kernel_mode_supported) supported;
    decltype(&simpler_kernel_mode_init) init;
    decltype(&simpler_kernel_mode_prepare_callable) prepare_callable;
    decltype(&simpler_kernel_mode_launch) launch;
    decltype(&finalize_device) finalize;
};

template <typename F>
F resolve(void *handle, const char *name) {
    dlerror();
    void *sym = dlsym(handle, name);
    if (dlerror() != nullptr) return nullptr;
    return reinterpret_cast<F>(sym);
}

bool load_kernel_mode_api(void *handle, KernelModeApi &api) {
    api.create_device_context = resolve<decltype(api.create_device_context)>(handle, "create_device_context");
    api.destroy_device_context = resolve<decltype(api.destroy_device_context)>(handle, "destroy_device_context");
    api.supported = resolve<decltype(api.supported)>(handle, "simpler_kernel_mode_supported");
    api.init = resolve<decltype(api.init)>(handle, "simpler_kernel_mode_init");
    api.prepare_callable = resolve<decltype(api.prepare_callable)>(handle, "simpler_kernel_mode_prepare_callable");
    api.launch = resolve<decltype(api.launch)>(handle, "simpler_kernel_mode_launch");
    api.finalize = resolve<decltype(api.finalize)>(handle, "finalize_device");
    return api.create_device_context && api.destroy_device_context && api.supported && api.init &&
           api.prepare_callable && api.launch && api.finalize;
}

std::vector<uint8_t> read_binary(const char *path) {
    std::ifstream stream(path, std::ios::binary);
    if (!stream) throw std::runtime_error(std::string("cannot read runtime binary: ") + path);
    return {std::istreambuf_iterator<char>(stream), std::istreambuf_iterator<char>()};
}

/// Device ids allocated by CTest resource allocation, parsed from
/// CTEST_RESOURCE_GROUP_<n>_NPUS ("id:<N>,slots:<M>;..."). Mirrors
/// test_comm_lifecycle.cpp::read_ctest_devices().
std::vector<int> read_ctest_devices() {
    std::vector<int> ids;
    const char *count_str = std::getenv("CTEST_RESOURCE_GROUP_COUNT");
    if (count_str == nullptr) return ids;

    int group_count = std::atoi(count_str);
    for (int g = 0; g < group_count; ++g) {
        std::string var = "CTEST_RESOURCE_GROUP_" + std::to_string(g) + "_NPUS";
        const char *val = std::getenv(var.c_str());
        if (val == nullptr) continue;

        std::string s(val);
        size_t pos = 0;
        while ((pos = s.find("id:", pos)) != std::string::npos) {
            pos += 3;
            ids.push_back(std::atoi(s.c_str() + pos));
        }
    }
    return ids;
}

/// A correctly aligned header used only for malformed-input and pre-init tests.
struct AlignedCallableImage {
    alignas(ChipCallable) unsigned char bytes[sizeof(ChipCallable)];

    AlignedCallableImage() { std::memset(bytes, 0, sizeof(bytes)); }
    const void *data() const { return bytes; }
    static size_t size() { return sizeof(ChipCallable); }
};

/// The caller's device and stream, stood up and torn down by the test. This is
/// the state a kernel-mode caller already holds and lends to simpler; nothing
/// under test may reset or finalize any of it.
class BorrowedDevice {
public:
    explicit BorrowedDevice(int device_id) :
        device_id_(device_id) {
        aclError rc = aclInit(nullptr);
        // aclInit is process-wide and another owner in this process may have
        // run it already.
        acl_owned_ = (rc == ACL_SUCCESS);
        if (rc != ACL_SUCCESS && static_cast<int>(rc) != ACL_ERROR_REPEAT_INITIALIZE) return;
        if (aclrtSetDevice(device_id_) != ACL_SUCCESS) return;
        device_set_ = true;
        if (aclrtCreateStream(&stream_) != ACL_SUCCESS) stream_ = nullptr;
    }

    ~BorrowedDevice() {
        if (stream_ != nullptr) aclrtDestroyStream(stream_);
        if (device_set_) aclrtResetDevice(device_id_);
        if (acl_owned_) aclFinalize();
    }

    BorrowedDevice(const BorrowedDevice &) = delete;
    BorrowedDevice &operator=(const BorrowedDevice &) = delete;

    bool ready() const { return device_set_ && stream_ != nullptr; }
    void *stream() const { return stream_; }
    int device_id() const { return device_id_; }

private:
    int device_id_;
    bool acl_owned_{false};
    bool device_set_{false};
    aclrtStream stream_{nullptr};
};

class KernelModeEntryTest : public ::testing::Test {
protected:
    void SetUp() override {
        for (const char *path :
             {PTO_HOST_RUNTIME_LIB_PATH, PTO_KERNEL_UT_AICPU_PATH, PTO_KERNEL_UT_AICORE_PATH,
              PTO_KERNEL_UT_DISPATCHER_PATH}) {
            if (!std::filesystem::exists(path)) {
                GTEST_SKIP() << "onboard runtime artifact not built: " << path
                             << "\n(build the a2a3 onboard tensormap_and_ringbuffer runtime first)";
            }
        }
        auto devices = read_ctest_devices();
        if (devices.empty()) {
            GTEST_SKIP() << "no NPU device allocated; run ctest with --resource-spec-file";
        }
        device_id_ = devices[0];
    }

    int device_id_{-1};
};

// ---------------------------------------------------------------------------
// C ABI surface: all four entries, on a borrowed device and stream.
// ---------------------------------------------------------------------------

TEST_F(KernelModeEntryTest, InitAndFinalizePreserveBorrowedStream) {
    BorrowedDevice borrowed(device_id_);
    ASSERT_TRUE(borrowed.ready()) << "could not stand up the caller's device/stream on device " << device_id_;

    void *handle = dlopen(PTO_HOST_RUNTIME_LIB_PATH, RTLD_NOW | RTLD_LOCAL);
    ASSERT_NE(handle, nullptr) << "dlopen failed: " << dlerror();

    KernelModeApi api{};
    ASSERT_TRUE(load_kernel_mode_api(handle, api)) << "kernel-mode entries missing from libhost_runtime.so";

    void *ctx = api.create_device_context();
    ASSERT_NE(ctx, nullptr);

    EXPECT_EQ(api.supported(ctx), 0) << "this backend reports kernel mode unsupported";

    const CallConfig config{};
    const auto aicpu = read_binary(PTO_KERNEL_UT_AICPU_PATH);
    const auto aicore = read_binary(PTO_KERNEL_UT_AICORE_PATH);
    const auto dispatcher = read_binary(PTO_KERNEL_UT_DISPATCHER_PATH);

    EXPECT_EQ(
        api.init(
            ctx, device_id_, aicpu.data(), aicpu.size(), aicore.data(), aicore.size(), dispatcher.data(),
            dispatcher.size(), &config, 1
        ),
        0
    );
    EXPECT_EQ(api.supported(ctx), 0);

    ChipStorageTaskArgs args{};
    EXPECT_EQ(api.launch(ctx, 0, &args, borrowed.stream()), PTO_RUNTIME_ERR_INVALID_STATE)
        << "execution providers are not bound";

    EXPECT_EQ(api.finalize(ctx), 0);
    api.destroy_device_context(ctx);
    dlclose(handle);
    EXPECT_EQ(aclrtSynchronizeStream(borrowed.stream()), ACL_SUCCESS);
}

TEST_F(KernelModeEntryTest, MalformedArgumentsAreRejectedBeforeTheStateCheck) {
    BorrowedDevice borrowed(device_id_);
    ASSERT_TRUE(borrowed.ready());

    void *handle = dlopen(PTO_HOST_RUNTIME_LIB_PATH, RTLD_NOW | RTLD_LOCAL);
    ASSERT_NE(handle, nullptr) << "dlopen failed: " << dlerror();
    KernelModeApi api{};
    ASSERT_TRUE(load_kernel_mode_api(handle, api));
    void *ctx = api.create_device_context();
    ASSERT_NE(ctx, nullptr);

    const CallConfig config{};
    const std::vector<uint8_t> aicpu(16, 0);

    // Generation zero is the ABI's invalid generation.
    EXPECT_EQ(
        api.init(ctx, device_id_, aicpu.data(), aicpu.size(), nullptr, 0, nullptr, 0, &config, 0),
        PTO_RUNTIME_ERR_INTERNAL
    );
    // A binary and its size describe one object: present-together or absent-together.
    EXPECT_EQ(api.init(ctx, device_id_, aicpu.data(), 0, nullptr, 0, nullptr, 0, &config, 1), PTO_RUNTIME_ERR_INTERNAL);
    EXPECT_EQ(
        api.init(ctx, -1, aicpu.data(), aicpu.size(), nullptr, 0, nullptr, 0, &config, 1), PTO_RUNTIME_ERR_INTERNAL
    );

    AlignedCallableImage image;
    EXPECT_EQ(api.prepare_callable(ctx, 0, nullptr, AlignedCallableImage::size()), PTO_RUNTIME_ERR_INTERNAL);
    // Below the size floor the image cannot hold a ChipCallable header.
    EXPECT_EQ(api.prepare_callable(ctx, 0, image.data(), sizeof(ChipCallable) - 1), PTO_RUNTIME_ERR_INTERNAL);
    EXPECT_EQ(
        api.prepare_callable(ctx, MAX_REGISTERED_CALLABLE_IDS, image.data(), AlignedCallableImage::size()),
        PTO_RUNTIME_ERR_INTERNAL
    );

    ChipStorageTaskArgs args{};
    EXPECT_EQ(api.launch(ctx, 0, &args, nullptr), PTO_RUNTIME_ERR_INTERNAL);
    EXPECT_EQ(api.launch(ctx, -1, &args, borrowed.stream()), PTO_RUNTIME_ERR_INTERNAL);

    api.destroy_device_context(ctx);
    dlclose(handle);
}

// ---------------------------------------------------------------------------
// ChipWorker surface: the host half of the wiring.
// ---------------------------------------------------------------------------

TEST_F(KernelModeEntryTest, WorkerKernelInitAndClosePreserveBorrowedStream) {
    BorrowedDevice borrowed(device_id_);
    ASSERT_TRUE(borrowed.ready());

    ChipWorker worker;
    const CallConfig config{};
    ASSERT_NO_THROW(worker.kernel_init(
        PTO_HOST_RUNTIME_LIB_PATH, PTO_KERNEL_UT_AICPU_PATH, PTO_KERNEL_UT_AICORE_PATH, PTO_KERNEL_UT_DISPATCHER_PATH,
        device_id_, config, ChipWorker::next_kernel_context_generation()
    ));
    EXPECT_TRUE(worker.initialized());
    EXPECT_EQ(worker.device_id(), device_id_);
    EXPECT_FALSE(worker.kernel_mode_supported());
    ChipStorageTaskArgs args{};
    try {
        worker.kernel_launch(0, &args, borrowed.stream());
        FAIL() << "launch must reject missing execution providers";
    } catch (const std::runtime_error &error) {
        EXPECT_NE(std::string(error.what()).find(std::to_string(PTO_RUNTIME_ERR_INVALID_STATE)), std::string::npos);
    }
    EXPECT_NO_THROW(worker.finalize());
    EXPECT_FALSE(worker.initialized());
    EXPECT_EQ(aclrtSynchronizeStream(borrowed.stream()), ACL_SUCCESS);
}

TEST_F(KernelModeEntryTest, WorkerKernelStateMachineRejectsOutOfOrderUse) {
    BorrowedDevice borrowed(device_id_);
    ASSERT_TRUE(borrowed.ready());

    const CallConfig config{};
    AlignedCallableImage image;
    ChipStorageTaskArgs args{};

    // Before any init there is no runtime to reach, so the host half refuses.
    {
        ChipWorker worker;
        EXPECT_THROW(worker.kernel_mode_supported(), std::runtime_error);
        EXPECT_THROW(worker.kernel_prepare_callable(0, image.data(), AlignedCallableImage::size()), std::runtime_error);
        EXPECT_THROW(worker.kernel_launch(0, &args, borrowed.stream()), std::runtime_error);
    }

    // A null stream never reaches the runtime.
    {
        ChipWorker worker;
        EXPECT_THROW(worker.kernel_launch(0, &args, nullptr), std::runtime_error);
    }

    // Generation zero is refused on the host side, before the C ABI would.
    {
        ChipWorker worker;
        EXPECT_THROW(
            worker.kernel_init(
                PTO_HOST_RUNTIME_LIB_PATH, PTO_KERNEL_UT_AICPU_PATH, PTO_KERNEL_UT_AICORE_PATH,
                PTO_KERNEL_UT_DISPATCHER_PATH, device_id_, config, 0
            ),
            std::invalid_argument
        );
    }

    // After finalize the worker is terminal for both identities.
    {
        ChipWorker worker;
        worker.finalize();
        EXPECT_THROW(
            worker.kernel_init(
                PTO_HOST_RUNTIME_LIB_PATH, PTO_KERNEL_UT_AICPU_PATH, PTO_KERNEL_UT_AICORE_PATH,
                PTO_KERNEL_UT_DISPATCHER_PATH, device_id_, config, ChipWorker::next_kernel_context_generation()
            ),
            std::runtime_error
        );
    }
}

TEST_F(KernelModeEntryTest, ContextGenerationsAreNonzeroAndIncreasing) {
    const uint64_t first = ChipWorker::next_kernel_context_generation();
    const uint64_t second = ChipWorker::next_kernel_context_generation();
    EXPECT_NE(first, 0u) << "generation zero is what the C ABI rejects as invalid";
    EXPECT_GT(second, first);
}

}  // namespace
